src/citrine/informatics/data_sources.py

```diff
@@ -0,0 +1,99 @@
+"""Tools for working with Descriptors."""
+from abc import abstractmethod
+from typing import Type, List, Mapping  # noqa: F401
+from uuid import UUID
+
+from citrine._serialization.serializable import Serializable
+from citrine._serialization.polymorphic_serializable import PolymorphicSerializable
+from citrine._serialization import properties
+from citrine.informatics.descriptors import Descriptor
+from citrine.resources.file_link import FileLink
+
+
+class DataSource(PolymorphicSerializable['DataSource']):
+    """[ALPHA] A source of data for the AI engine.
+
+    Data source provides a polymorphic interface for specifying different kinds of data as the
+    training data for predictors and the input data for some design spaces.
+
+    """
+
+    @abstractmethod
+    def _attrs(self) -> List[str]:
+        pass  # pragma: no cover
+
+    def __eq__(self, other):
+        try:
+            return all([
+                self.__getattribute__(key) == other.__getattribute__(key) for key in self._attrs()
+            ])
+        except AttributeError:
+            return False
+
+    @classmethod
+    def get_type(cls, data) -> Type[Serializable]:
+        """Return the subtype."""
+        if "type" not in data:
+            raise ValueError("Can only get types from dicts with a 'type' key")
+        types: List[Type[Serializable]] = [
+            CSVDataSource, AraTableDataSource
+        ]
+        res = next((x for x in types if x.type == data["type"]), None)
+        if res is None:
+            raise ValueError("Unrecognized type: {}".format(data["type"]))
+        return res
+
+
+class CSVDataSource(Serializable['CSVDataSource'], DataSource):
+    """[ALPHA] A data source based on a CSV file stored on the data platform.
+
+    Parameters
+    ----------
+    file_link: FileLink
+        link to the CSV file to read the data from
+    column_definitions: Mapping[str, Descriptor]
+        Map the column headers to the descriptors that will be used to interpret the cell contents
+
+    """
+
+    type = properties.String('type', default='csv_data_source', deserializable=False)
```

Comment:
@talltea note

Reply:
Haven't we been naming this variable `typ`?

Reply:
I copied the source file from `descriptors.py` as a starting point, and [that used `type`](https://github.com/CitrineInformatics/citrine-python/blob/master/src/citrine/informatics/descriptors.py#L44).  I actually prefer `typ`; will change.

Reply:
It looks like the rest of the `informatics` module uses `typ` too, so I'm going to change it in `descriptors.py` too.